pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -103,13 +103,14 @@ public void addSegment(IndexSegment segment, Iterator<RecordInfo> recordInfoIter
           (primaryKey, currentRecordLocation) -> {
             if (currentRecordLocation != null) {
               // Existing primary key
+              IndexSegment currentSegment = currentRecordLocation.getSegment();
+              int comparisonResult = recordInfo._comparisonValue.compareTo(currentRecordLocation.getComparisonValue());
```

Comment:
curious, why particular line caused the bug?

Reply:
line 140 in the old code (using `==` instead of `equals()` for boxed value)

Reply:
I see. makes sense.